### PR TITLE
Solve (1/x - 1).is_nonpositive should be True if x is positive integer

### LINKED ISSUE #10458: [enhancement] (1/x - 1).is_nonpositive should be True if x is positive integer
```python
>>> var('x', integer=True, positive=True)
x
>>> ask(Q.nonpositive(1/x-1))
>>> (1/x-1).is_nonpositive
```

Not sure of the best test, but this should be able to return True.

### PR DESCRIPTION
**Pull request by mvnnn**
fixes #10458

### CODE REVIEW

**Comment by mvnnn** [comment]
below test case Failed ::
```
line 1518, in test_negative   assert ask(Q.negative(cos(I)**2 + sin(I)**2 - 1)) is None
line 1566, in test_nonzeroassert ask(Q.nonzero(cos(1)**2 + sin(1)**2 - 1)) is None

**Comment by asmeurer** [comment]
simplify is a potentially expensive function call. Adding it to ask could make ask potentially very slow.

**Comment by mvnnn** [comment]
then what should i do for fix this ?

**Comment by asmeurer** [comment]
You need to add logic to the assumptions system so that it can deduct `((1/x) - 1).is_nonpositive`.  I'd have to think about how to go about doing that. The key reason it's nonpositive is that x is an integer, which means that it's >=1, so that 1/x is <=1. I don't know the best way to do that, but I would start by figuring out why `(x - 1).is_nonpositive` works.

**Comment by smichr** [comment]
There are a few special cases that can be handled and I think some of the other `is_` routines implement them so you could look at them to see how it is done. Maybe ones related to being prime? Anyway, we have a few special integers: 1 which is the smallest positive integer and 2 which is the smallest even integer, so var('i', even=True >= 2` and `var('i', integer=True, positive=True) >= 1`.

**Comment by mvnnn** [comment]
@asmeurer i agree simplify make ask potentially very slow but when we use ```sympify```
```
>>> sympify('sin(x)**2 + cos(x)**2')
sin(x)**2 + cos(x)**2
>>> sympify('1/x - 1')
-1 + 1/x
```
because sympify have no methods for trigonometry formula  and many other possible methods and it's can't convert ```1/x-1``` to ```(1-x)/x```. simplify having only those method which doing adding and subtracting ..
but in case of ```simplify``` 
```
>>> simplify('1/x - 1')
(-x + 1)/x
>>> simplify('sin(x)**2 + cos(x)**2')
1
```

**Comment by mvnnn** [comment]
another option is that to add those functionality in simpify.

**Comment by asmeurer** [comment]
Can you explain your latest change here? Your commit message isn't very helpful.

**Comment by mvnnn** [comment]
now it's output ::
```
>>> (1/x-1).is_nonpositive
True
>>> (1-1/x).is_nonpositive
>>> (1-1/x).is_nonnegative
True
>>> (1/x-1).is_nonnegative
```

File: sympy/assumptions/ask.py [hunk 1]
```diff
@@ -1253,9 +1253,9 @@ def ask(proposition, assumptions=True, context=global_assumptions):
         raise TypeError("assumptions must be a valid logical expression")
 
     if isinstance(proposition, AppliedPredicate):
-        key, expr = proposition.func, sympify(proposition.arg)
+        key, expr = proposition.func, proposition.arg
     else:
-        key, expr = Q.is_true, sympify(proposition)
```
Comment (asmeurer) [inline sympy/assumptions/ask.py]: I'm assuming this was removed on accident.

**Comment by asmeurer** [comment]
I'm still -1 to calling simplify in the assumptions.

**Comment by asmeurer** [comment]
Also, don't add `[skip ci]` to your commit message. That will prevent Travis from running the tests.

**Comment by asmeurer** [comment]
(By the way, I think your system clock is wrong. The timestamp on your commit is `Fri Jan 29 03:04:42 2016 +0530` even though the current time is `Thu Jan 28 9:58:00 2016 +0530`. I noticed because it always appears below new comments that I make here on this pull request)

**Comment by mvnnn** [comment]
```
>>> (sin('x')**2 + cos('x')**2 + x).is_nonnegative
True
>>> (sin('x')**2 + cos('x')**2 - x).is_nonnegative
```
for ```trigonometry formula``` sympify have no method to solve this so output is ```None```but ```simplify``` have method for this so it's give correct output.

File: sympy/core/add.py [hunk 2]
```diff
@@ -593,9 +593,11 @@ def _eval_is_positive(self):
 
     def _eval_is_nonnegative(self):
         from sympy.core.exprtools import _monotonic_sign
+        from sympy import simplify
+        self = simplify(self)
         if not self.is_number:
             c, a = self.as_coeff_Add()
-            if not c.is_zero and a.is_nonnegative:
```
Comment (asmeurer) [inline sympy/core/add.py]: Why did you change this?

Comment (mvnnn) [inline sympy/core/add.py]: ```(1/x-1)``` in this case simplify convert it to ```(1-x)/x``` so this case ```c=0```

Comment (asmeurer) [inline sympy/core/add.py]: This code is assuming that `self` is an Add. The above call to simplify breaks that.

Comment (mvnnn) [inline sympy/core/add.py]: ```as_coeff_Add()``` function return ```1/x``` value--->``` a  = _pos``` and ```x``` value---> ```a =x```.
if we make condition if ```a = variable``` then we not need to use simplify and if ```a=_pos/_neg``` in that case we use simplify.it's solve also ```trigonometry formula``` as problem in above example.

Comment (mvnnn) [inline sympy/core/add.py]: other option is that to modify ```sympify``` functionality.

**Comment by asmeurer** [comment]
@mvnnn I encourage you to run the tests `./bin/test` so you can see how this is breaking things. And in the future, don't add "[skip ci]" to your commit messages, so that Travis will run the tests too.

**Comment by smichr** [comment]
I affirmt that the use of simplify in assumptions is not a good idea. That is the nature of the problem: if the assumptions can't answer they return None and it is up to the user to figure out if simplification is needed to get the expression into shape.

The use of `sympify` is to make sure that something like `True` gets converted to `S.true`, I suspect. It is not intended to do re-writing.

**Comment by mvnnn** [comment]
@asmeurer @smichr thanks for review, i will try to fix without ```simplify```

**Comment by mvnnn** [comment]
Does this look good.@asmeurer@smichr

File: sympy/core/add.py [hunk 3]
```diff
@@ -610,10 +612,12 @@ def _eval_is_nonpositive(self):
         from sympy.core.exprtools import _monotonic_sign
         if not self.is_number:
             c, a = self.as_coeff_Add()
-            if not c.is_zero and a.is_nonpositive:
+            if not c.is_zero:
                 v = _monotonic_sign(a)
                 if v is not None:
                     s = v + c
+                    if a.is_nonnegative and v.is_integer == True:
```
Comment (asmeurer) [inline sympy/core/add.py]: Can you explain why this change works? I'm unclear what this code actually does (it could use some comments) and why is_integer would come into play here.

**Comment by mvnnn** [comment]
Does this look good.@asmeurer@smichr

File: sympy/core/add.py (hunk 3 above)
Comment (mvnnn) [inline sympy/core/add.py]: is_integer can make difference between "x" and "1/x"

Comment (mvnnn) [inline sympy/core/add.py]: @asmeurer i will add comments and more complicated test cases for ```is_nonnegative``` and ```is_nonpositive```. should i do ?

Comment (asmeurer) [inline sympy/core/add.py]: Yes

Comment (mvnnn) [inline sympy/core/add.py]: @asmeurer Does this look good ?

File: sympy/assumptions/tests/test_query.py [hunk 4]
```diff
@@ -2186,6 +2186,32 @@ def test_issue_9636():
...
+
+    x = symbols('x', integer=True, negative=True)
+    assert ask(Q.nonpositive(1/x-1)) is True
+    assert ask(Q.nonpositive(1-1/x)) is False
+    assert ask(Q.nonnegative(1/x-1)) is False
+    assert ask(Q.nonnegative(1-1/x)) is True
+    assert ask(Q.nonpositive(-x+x**2+1)) is False
+    assert ask(Q.nonnegative(-x+x**2+1)) is True
+    assert ask(Q.nonpositive(1-(1/x**2)-x**2+3*x)) is True
+    assert ask(Q.nonnegative(1-(1/x**2)-x**2+3*x)) is False
```
Comment (debugger22) [inline sympy/assumptions/tests/test_query.py]: It's a good idea to use whitespaces with `-` and `+` operators.

File: sympy/core/add.py [hunk 5]
```diff
@@ -595,10 +595,13 @@ def _eval_is_nonnegative(self):
         from sympy.core.exprtools import _monotonic_sign
         if not self.is_number:
             c, a = self.as_coeff_Add()
-            if not c.is_zero and a.is_nonnegative:
+            if not c.is_zero:
                 v = _monotonic_sign(a)
                 if v is not None:
                     s = v + c
+                    # make difference between integer and rational number
+                    if a.is_nonpositive and v.is_integer == True:
```
Comment (debugger22) [inline sympy/core/add.py]: You can remove equality here.

File: sympy/functions/elementary/tests/test_exponential.py [hunk 6]
```diff
@@ -381,7 +381,7 @@ def test_lambertw():
     assert LambertW(2, evaluate=False).is_real
     p = Symbol('p', positive=True)
     assert LambertW(p, evaluate=False).is_real
-    assert LambertW(p - 1, evaluate=False).is_real is None
+    assert LambertW(p - 1, evaluate=False).is_real is False
```
Comment (debugger22) [inline sympy/functions/elementary/tests/test_exponential.py]: I'm not sure about this change. @asmeurer Is this correct?

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: I think the correct answer should be True http://www.wolframalpha.com/input/?i=lambertw&dataset=

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: And even the test above this one shows that it can't be False.

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: ```p``` is positive but not ```integer``` so answer is ```false```.

Comment (debugger22) [inline sympy/functions/elementary/tests/test_exponential.py]: It can not be `False`. It can be `None` but not `False`.

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: at line 380 ```LambertW(-1).is_real is False```

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: I guess LambertW is non-real for negative values greater than -1 (like -1/2). So this should be None.

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: I guess the cutoff point is -1/E https://en.wikipedia.org/wiki/Lambert_W_function.

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: Oh, I'll fix that

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: when we call ```LambertW(p - 1, evaluate=False).is_real```--->
https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/exponential.py#L842
--->https://github.com/sympy/sympy/blob/master/sympy/core/add.py#L612 this function give wrong answer ```c = -1 and a = p + exp(-1)``` but right answer is ```c = -1 + exp(-1) and a = p``` so we need to modify ```as_coeff_Add()``` function.

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: after doing this ```c = -1 + exp(-1)``` we need to convert c into ```(-E + 1)*exp(-1)```

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: other bad way is pass value i mean pass ```1/e = 0.3679```
https://github.com/sympy/sympy/blob/master/sympy/functions/elementary/exponential.py#L842

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: That's how as_coeff_Add works. The first term is always an integer or rational number.

File: sympy/core/add.py [hunk 7]
```diff
@@ -595,10 +595,13 @@ def _eval_is_nonnegative(self):
         from sympy.core.exprtools import _monotonic_sign
         if not self.is_number:
             c, a = self.as_coeff_Add()
-            if not c.is_zero and a.is_nonnegative:
+            if not c.is_zero:
                 v = _monotonic_sign(a)
```
Comment (asmeurer) [inline sympy/core/add.py]: It looks like the bulk of the logic here is done in this function.

File: sympy/functions/elementary/tests/test_exponential.py (hunk 6 above)
Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: i mean ```1/e``` is constant not variable so we need to add them in to ```c```(constant)

File: sympy/core/add.py (hunk 7 above)
Comment (mvnnn) [inline sympy/core/add.py]: but it can't allow to pass ```(1 - 1/x)``` --> c=1 and a = "neg".

File: sympy/functions/elementary/tests/test_exponential.py (hunk 6 above)
Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: if i declare ```e```value in this class then it's solve this problem.should i do ?

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: I'm unclear what you are suggesting to do.

Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: ```sympy/sympy/core/add.py``` in this class if we declare ```e=2.71```then we can solve above problem.

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: Using floats over symbolic expressions is generally not a good idea, as it can lead to subtle issues. For example, what if 2.71 is not enough precision for some expression. What if someone passes in `exp(1)` to the function? `(exp(1) - 2.71).is_positive` will be True, even though the intended effect will be for it to be True.

File: sympy/core/add.py (hunk 7 above)
Comment (asmeurer) [inline sympy/core/add.py]: I don't understand why c is pulled out before passing the expression to this function. What happens if you just use `_monotonic_sign(self)`?

Comment (mvnnn) [inline sympy/core/add.py]: ```_monotonic_sign(self)``` in this case it's return wrong value.

Comment (mvnnn) [inline sympy/core/add.py]: i think ```abs()``` is working correct in this case.

Comment (mvnnn) [inline sympy/core/add.py]: it's return ```None``` but right value is ```True```(in sympy live shell)
```
>>> p = Symbol('p', positive=True)
>>> LambertW(p - 2/S.Exp1 + S.Exp1, evaluate=False).is_real
```

Comment (mvnnn) [inline sympy/core/add.py]: when in expression S.Exp() include then we take abs(S.Exp()) and we declare in that class e value is 2.71828182846 {abs(e)=2.71828182846}.

Comment (mvnnn) [inline sympy/core/add.py]: i want to modify ```as_coeff_Add()``` function. i make changes is ```c = -1``` and ```a = p + exp(-1)``` to ```c = -1 + exp(-1)``` and ```a = p```. i mean currently ```as_coeff_Add()``` include ```exp``` in ```a``` part and i change to ```exp``` include in ```c``` part.@asmeurer should i do ?

Comment (asmeurer) [inline sympy/core/add.py]: No, as_coeff_Add should only extract the purely numeric (i.e., integer, rational, or float) term. I would focus on fixing the _monotonic_sign function, and understanding why we can't just use _monotonic_sign(self).

File: sympy/functions/elementary/tests/test_exponential.py (hunk 6 above)
Comment (smichr) [inline sympy/functions/elementary/tests/test_exponential.py]: > That's how as_coeff_Add works. The first term is always an integer or rational number.

You can use `as_independent`, however:

```
>>> eq=1+exp(1)+x+y
>>> eq.as_independent(*eq.free_symbols, as_Add=True)
(1 + E, x + y)
```

Comment (asmeurer) [inline sympy/functions/elementary/tests/test_exponential.py]: I still don't understand why it's needed at all, though. The docstring of `_monotonic_sign` suggests to me that it should work on the whole expression.

**Comment by smichr** [comment]
I have sent some [modifications](https://github.com/mvnnn/sympy/pull/1).

File: sympy/functions/elementary/tests/test_exponential.py (hunk 6 above)
Comment (mvnnn) [inline sympy/functions/elementary/tests/test_exponential.py]: @smichr ```eq.as_independent``` working good but sometime ```is_nonpositive``` return ```None``` ans and sometime return ```correct``` ans  for same input.
```
>>> x = Symbol('x', integer=True, negative=True)
>>> (1 - 1/x).is_nonpositive
```
it's return ```None```
```
>>> x = Symbol('x', integer=True, negative=True)
>>> (1 - 1/x).is_nonnegative
True
>>> (1 - 1/x).is_nonpositive
False
```

**Comment by czgdp1807** [comment]
@mvnnn Would you like to continue this PR? Its a good patch. Thanks for the contribution.

**Comment by czgdp1807** [comment]
Due to inactivity, I am closing this for now. However, please let us know if someone wants to continue this as the issue is still unresolved and is open. Thanks for the contributions and apologies for inconveniences caused.
